### PR TITLE
Move load_state_dict into LitModular

### PR DESCRIPTION
**Pull request by dxoigmn**
# What does this PR do?

The mechanism for wrapping a module in `mart.nn.load_state_dict` makes YAML configuration more verbose and is unnecessary if that logic is moved into `LitModular`. Ideally, however, we would support custom syntax within the YAML configuration for things that do not support weight loading. For example:
```yaml
model:
  _target_: torchvision.models.detection.fasterrcnn_resnet50_fpn
  _load_state_dict_: path/to/weights.pt
```

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

## Before submitting

- [ ] The title is **self-explanatory** and the description **concisely** explains the PR
- [ ] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [ ] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/models/modular.py [hunk 1]
```diff
@@ -80,6 +78,13 @@ def __init__(
         self.test_step_log = test_step_log or []
         self.test_metrics = test_metrics
 
+        # Load state dict for specified modules
+        load_state_dict = load_state_dict or {}
+        for name, path in load_state_dict.items():
+            module = attrgetter(name)(self.model)
```
Comment (mzweilin) [inline mart/models/modular.py:84]: Though it gets nested attributes now, I'm not sure how to add a name with dots in the Hydra command  line.

Comment (dxoigmn) [reply mart/models/modular.py:84]: Ah `attrgetter` is a nice find! Looks like we'd have to change the separator for it to work on the commandline.

Comment (mzweilin) [reply mart/models/modular.py:86]: I added a `get_dottedpath_items()` to extract dotted paths from nested dictionaries. So we can keep the command line tiny.

**Comment by mzweilin** [comment]
Usage example:

```bash
python -m mart \
experiment=ArmoryCarlaOverObjDet_TorchvisionFasterRCNN \
fit=false \
+trainer.limit_test_batches=1 \
+model.load_state_dict.losses_and_detections.model=/home/weilinxu/coder/GARD-with-MART/oscar/model_zoo/carla_rgb_weights_eval6.pt
```

File: mart/models/modular.py [hunk 2]
```diff
@@ -80,6 +79,14 @@ def __init__(
         self.test_step_log = test_step_log or []
         self.test_metrics = test_metrics
 
+        # Load state dict for specified modules
+        load_state_dict = load_state_dict or {}
+        # Hydra commandlines converts dotted path to nested dictionary.
+        for name, path in get_dottedpath_items(load_state_dict):
+            module = attrgetter(name)(self.model)
+            logger.info(f"Loading state_dict {path} for {module.__class__.__name__}...")
+            module.load_state_dict(torch.load(path))
```
Comment (mzweilin) [inline mart/models/modular.py:88]: Shall we add `map_location="cpu"` in case no GPU is available?

Comment (dxoigmn) [reply mart/models/modular.py:88]: I guess that is what is in the original implementation so sure.

File: mart/models/modular.py (hunk 1 above)
Comment (dxoigmn) [reply mart/models/modular.py:86]: Can you give me an example command line? Trying to understand the code you wrote.

**Comment by dxoigmn** [comment]
What was the previous command-line for loading this checkpoint?

File: mart/models/modular.py (hunk 1 above)
Comment (dxoigmn) [reply mart/models/modular.py:86]: Okay now I get what you implemented: flatten. Can you rename this to flatten. One issue, however, is that your implementation will not fail in the case of two keys with the same name. For example:
```json
{
  "logits": { "0": "0.pt" },
  "logits.0": "logits_0.pt",
}
```

I feel like that should throw an error.

Comment (mzweilin) [reply mart/models/modular.py:86]: I found a package https://flatdict.readthedocs.io/en/stable/#docs
, but it does not detect key collision either.

```python
>>> from flatdict import FlatDict
>>> FlatDict({"a": {"b": "c.pth"}}, delimiter=".")["a.b"]
'c.pth'
>>> FlatDict({"a": {"b": "c.pth"}, "a.b": "d.pth"}, delimiter=".")["a.b"]
'd.pth'
```

Comment (mzweilin) [reply mart/models/modular.py:86]: `pandas` does not detect key collision either.

```python
>>> import pandas as pd

>>> d = {'a': 1,
     'c.a': 1,
     'c': {'a': 2, 'b': {'x': 5, 'y' : 10}},
     'd': [1, 2, 3]}

>>> df = pd.json_normalize(d, sep='.')

>>> print(df.to_dict(orient='records')[0])
{'a': 1, 'c.a': 2, 'd': [1, 2, 3], 'c.b.x': 5, 'c.b.y': 10}
```

Comment (mzweilin) [reply mart/models/modular.py:86]: I implemented `mart.utils.utils.FlattenDict` that detects key collision. Two test cases are added.

**Review by mzweilin — APPROVED** [review]
LGTM. Please take a look at the implementation of `mart.utils.utils.FlattenDict`.

File: mart/utils/utils.py [hunk 3]
```diff
@@ -272,3 +273,24 @@ def get_resume_checkpoint(config: DictConfig) -> Tuple[DictConfig]:
         config = hydra.compose(config_name, overrides=overrides)
 
     return config
+
+
+class FlattenDict(dict):
```
Comment (dxoigmn) [inline mart/utils/utils.py:278]: change to function

Comment (mzweilin) [reply mart/utils/utils.py:278]: done

**Review by mzweilin — APPROVED** [review]
LGTM. Please take a look at `mart.utils.flatten_dict`.

**Pull request merged**